client/my-sites/site-settings/site-icon-setting/index.jsx

```diff
@@ -76,22 +78,36 @@ class SiteIconSetting extends Component {
 			return;
 		}
 
+		// Upload media using a manually generated ID so that we can continue
+		// to reference it within this function
 		const transientMediaId = uniqueId( 'site-icon' );
-
 		MediaActions.add( siteId, {
 			ID: transientMediaId,
 			fileName: `cropped-${ selectedItem.file }`,
 			fileContents: blob
 		} );
 
+		// Avoid calling to component instance within progress check closure
+		// below since component may have been unmounted in time since upload
+		const {
+			saveSiteSettings: dispatchSaveSiteSettings,
+			setSiteIcon: dispatchSetSiteIcon
+		} = this.props;
```

Comment:
The Flux / Redux discrepancy makes this prohibitively difficult I think, since we want to optimistically update the site icon but not save the site settings until after the upload has completed.

If this were fully Redux, we might be able to achieve this within an action creator thunk and/or middleware. Otherwise we could consider some sort of sync mechanism like in #2167.

I think we may just be forced to bite the bullet and try to migrate media entirely to Redux, though for the time being I'd say that's outside the scope of what we're trying to achieve here.